File: Editorial/0008_String_to_Integer_atoi/C/1_Follow_the_Rules.c
int myAtoi(char* s) {
    long sign = 1;
    long result = 0;
    int i = 0;

    while (s[i] == ' ') {
        i++;
    }

    if (s[i] == '+') {
        sign = 1;
        i++;
    } else if (s[i] == '-') {
        sign = -1;
        i++;
    }

    while (s[i] != '\0' && s[i] >= '0' && s[i] <= '9') {
        if (result > INT_MAX / 10 ||
            (result == INT_MAX / 10 && s[i] - '0' > INT_MAX % 10)) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }
        result = result * 10 + (s[i++] - '0');
    }

    return sign * result;
}

File: Editorial/8_String_to_Integer_atoi/TypeScript/1_Follow_the_Rules.ts
function myAtoi(s: string): number {
    let sign = 1;
    let result = 0;
    let index = 0;
    let n = s.length;

    let INT_MAX = Math.pow(2, 31) - 1;
    let INT_MIN = -Math.pow(2, 31);

    while (index < n && s[index] == " ") {
        index++;
    }

    if (index < n && s[index] == "+") {
        sign = 1;
        index++;
    } else if (index < n && s[index] == "-") {
        sign = -1;
        index++;
    }

    while (index < n && s[index] >= "0" && s[index] <= "9") {
        let digit = Number(s[index]);

        if (
            result > Math.floor(INT_MAX / 10) ||
            (result == Math.floor(INT_MAX / 10) && digit > INT_MAX % 10)
        ) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }

        result = 10 * result + digit;
        index++;
    }

    return sign * result;
}
